src/gennylib/include/gennylib/PhaseLoop.hpp

```diff
@@ -77,10 +78,54 @@ class IterationCompletionCheck final {
         }
     }
 
-    explicit IterationCompletionCheck(PhaseContext& phaseContext)
-        : IterationCompletionCheck(phaseContext.get<TimeSpec, false>("Duration"),
-                                   phaseContext.get<UIntSpec, false>("Repeat"),
-                                   phaseContext.isNop()) {}
+    explicit IterationChecker(PhaseContext& phaseContext)
+        : IterationChecker(phaseContext.get<TimeSpec, false>("Duration"),
+                           phaseContext.get<IntegerSpec, false>("Repeat"),
+                           phaseContext.isNop()) {
+        auto rateSpec = phaseContext.get<RateSpec, false>("Rate");
+        const auto rateLimiterName =
+            phaseContext.get<std::string, false>("RateLimiterName").value_or("defaultRateLimiter");
+
+        if (rateSpec) {
+            if (!_doesBlock) {
+                throw InvalidConfigurationException(
+                    "Rate must be specified alongside either Duration or Repeat, otherwise there's "
+                    "no guarantee the rate limited operation will run in the correct phase");
+            }
+            _rateLimiter =
+                phaseContext.workload().getRateLimiter(rateLimiterName, rateSpec.value());
+        }
+    }
+
+    bool shouldLimitRate(int64_t currentIteration) const {
+        // Only rate limit if the current iteration is a muliple of the burst size.
+        return _rateLimiter && (currentIteration % _rateLimiter->getBurstSize() == 0);
```

Comment:
I'm going to push back again on this a bit (Sorry). I started changing `getBurstSize` to `mightExceedTokens` but didn't like the new interface. The problem is that it's now it mandates the caller to store `currentIteration`, which is not a thing that the caller needs and isn't coherent with the rest the token bucket class.

I image it should be perfectly fine to do something like:

```c++

if (rl.consume()) {
  for (i : rl.getBurstSize()) {
    // do something. 
  }
}
```

...Or the canonical token bucket example of slicing a burstSize() chunk of an array to be consumed. Neither of which require a notion of `currentIteration`.

Feel free to correct me if I'm wrong, but my understanding of your hesitation to exposing `consume()` and `burstSize()` is that these are implementation details of the rate limiter itself that should be exposed to the caller. A simpler interface would be exposing just `rateLimiter.doYourThing() # Blocking`. I would prefer this interface, but it makes the rate limiter's implementation a bit worse, because you'd need an atomic thing in the rate limiter to track the iteration, which creates a bottleneck.

If you prefer the simpler interface, I'd be happy to switch to the single call version of the rate limiter.

Reply:
ah so I'm fine with this as-is - why I said "consider" 😺 thanks for considering though seems like it's fine to leave I agree with your understanding.